vulnerabilities/data_source.py

```diff
@@ -319,35 +319,41 @@ def _collect_file_changes(
         file_ext: Optional[str],
     ) -> Tuple[Set[str], Set[str]]:
 
-        previous_commit = None
         added_files, updated_files = set(), set()
 
-        for commit in self._repo.walk(self._repo.head.target, pygit2.GIT_SORT_TIME):
-            commit_time = commit.commit_time + commit.commit_time_offset  # convert to UTC
-
-            if commit_time < self.cutoff_timestamp:
+        # find the most ancient commit we need to diff with
+        cutoff_commit = None
+        for commit in self._repo.iter_commits(self._repo.head):
+            if commit.committed_date < self.cutoff_timestamp:
                 break
+            cutoff_commit = commit
 
-            if previous_commit is None:
-                previous_commit = commit
-                continue
+        if cutoff_commit is None:
+            return added_files, updated_files
 
-            for d in commit.tree.diff_to_tree(previous_commit.tree).deltas:
-                if not _include_file(d.new_file.path, subdir, recursive, file_ext) or d.is_binary:
-                    continue
+        def _is_binary(d: DiffIndex):
+            if not d.b_blob:
+                return False
+            try:
+                d.b_blob.data_stream.read().decode()
+            except UnicodeDecodeError:
+                return True
+            return False
```

Comment:
What is this function trying to do ? Is it used to check whether the diff blob consists of binary ? This would return true if the diff is about something like a tar ball ?

Reply:
Indeed, it checks by trying to decode it with default encoding settings.

The previous code had some logic of checking if the data was binary to not count it as a diff.

The is_binary semantic was provided by libgit2. The logic is quite complicated:
https://github.com/libgit2/libgit2/blob/508361401fbb5d87118045eaeae3356a729131aa/src/buf_text.c#L268

Reply:
@sbs2001 as you suggested in the Gitter chat I guess we could use binaryornot which does a decent job there and is used extensively in scancode-toolkit:
- https://github.com/audreyfeldroy/binaryornot
- https://github.com/nexB/thirdparty-packages/blob/main/pypi/binaryornot-0.4.4.tar.gz.ABOUT
- https://github.com/nexB/typecode/blob/157b7be15c6cb4be754f915bde7974a3e0284961/src/typecode/contenttype.py#L926

Reply:
I changed the PR to use binaryornot. I am not sure why the nix CI is failing :-/